docs/async-await.md

```diff
@@ -0,0 +1,168 @@
+} catch {
+  print(error)
+}
+```
+
+If the library code throws from the AsyncRequest creation or the request invocation the user will, in normal use cases, handle the error in the same catch block. 
+
+#### Body streaming
+
+The new AsyncRequest has a new body type, that is wrapper around an internal enum. This allows us to evolve this type for use-cases that we are not aware of today. 
+
+```swift
+public struct Body {
+  static func bytes<S: Sequence>(_ sequence: S) -> Body where S.Element == UInt8
+
+  static func stream<S: AsyncSequence>(_ sequence: S) -> Body where S.Element == ByteBuffer
+
+  static func stream<S: AsyncSequence>(_ sequence: S) -> Body where S.Element == UInt8
+}
+```
+
+The main difference to today's `Request.Body` type is the lack of a `StreamWriter` for streaming scenarios. The existing StreamWriter offered the user an API to write into (thus the user was in control of when writing happened). The new `AsyncRequest.Body` uses `AsyncSequence`s to stream requests. By iterating over the provided AsyncSequence, the HTTPClient is  in control when writes happen, and can ask for more data efficiently. 
+
+Using the `AsyncSequence` from the Swift standard library as our upload stream mechanism dramatically reduces the learning curve for new users.
+
+### New `AsyncResponse` type
+
+The `AsyncResponse` looks more similar to the existing `Response` type. The biggest difference is again the `body` property, which is now an `AsyncSequence` of `ByteBuffer`s instead of a single optional `ByteBuffer?`. This will make every response on AsyncHTTPClient streaming by default.
+
+```swift
+public struct AsyncResponse {
+  /// the used http version
+  public var version: HTTPVersion
+  /// the http response status
+  public var status: HTTPResponseStatus
+  /// the response headers
+  public var headers: HTTPHeaders
+  /// the response payload as an AsyncSequence
+  public var body: Body
+}
+
+extension AsyncResponse {
+  public struct Body: AsyncSequence {
+    public typealias Element = ByteBuffer
+    public typealias AsyncIterator = Iterator
+  
+    public struct Iterator: AsyncIteratorProtocol {
+      public typealias Element = ByteBuffer
+    
+      public func next() async throws -> ByteBuffer?
+    }
+  
+    public func makeAsyncIterator() -> Iterator
+  }
+}
+```
+
+At a later point we could add trailers to the AsyncResponse as effectful properties.
+
+```swift
+    public var trailers: HTTPHeaders { async throws }
+```
+
+However we will need to make sure that the user has consumed the body stream completely before, calling the trailers, because otherwise we might run into a situation from which we can not progress forward:
+
+```swift
+do {
+  var request = AsyncRequest(url: "https://swift.org/")
+  let response = try await httpClient.execute(request, deadline: .now() + .seconds(3))
+  
+  var trailers = try await response.trailers // can not move forward since body must be consumed before.
```

Comment:
Probably throw if it is known that body is not consumed yet?

Reply:
I think we should just crash.

Reply:
I think I'm on the crash side as well. If the user consumes the trailers before the body, it will never be right. Crash seems to be the right tradeoff.

Reply:
Yeah that's true, unlikely it'll be sometimes correct, and even if it is people really must fix it. 👍